README.md
[![Build Status](https://travis-ci.org/binary-com/perl-YAML-CacheLoader.svg?branch=master)](https://travis-ci.org/binary-com/perl-YAML-CacheLoader)
[![Coverage Status](https://coveralls.io/repos/binary-com/perl-YAML-CacheLoader/badge.png?branch=master)](https://coveralls.io/r/binary-com/perl-YAML-CacheLoader?branch=master)
[![Gitter chat](https://badges.gitter.im/binary-com/perl-YAML-CacheLoader.png)](https://gitter.im/binary-com/perl-YAML-CacheLoader)

# FUNCTIONS

- LoadFile

    my $structure = LoadFile('/path/to/yml'\[, $force\_reload\]);

    Loads the structure from '/path/to/yml' into $structure, preferring the cached version if available,
    otherwise reading the file and caching the result for 593 seconds (about 10 minutes).

    If $force\_reload is set to a true value, the file will be loaded from disk without regard
    to the current cache status.

- DumpFile

    DumpFile('/path/to/yml', $structure);

    Dump the structure from $structure into '/path/to/yml', filling the cache along the way.

- FlushCache

    FlushCache();

    Remove all currently cached YAML documents from the cache server.

- FreshenCache

    FreshenCache();

    Freshen currently cached files which may be out of date, either by deleting the cache (for now deleted files) or reloading from the disk (for changed ones)

    May optionally provide a list of files to check, otherwise all known cached files are checked.

    Returns a stats hash-ref.


Changes
{{$NEXT}}

0.019     2017-02-27 07:14:15+00:00 UTC

0.018  2015-07-02 15:37:58 CAT
    - Remove FreshenCache from the general LoadFile flow

0.017  2015-03-02 11:55:51 MYT
    - Expand caching time to just over one hour from just under ten minutes

0.016  2015-01-31 12:35:48 CAT

    - Properly call YAML::XS

0.015  2015-01-31 10:58:53 CAT

    - Switch to less generic YAML::XS

0.014  2015-01-27 07:57:43 CAT

    - Fix Test::RedisServer dependency loading issue.

0.013  2015-01-21 14:35:46 CAT

    - Allow file list to FreshenCache
    - Use FreshenCache in the general LoadFile flow

0.012  2015-01-20 12:11:12 CAT

    - Require Cache::RedisDB 0.07
    - Improve FreshenCache with new 'ttl' command from Cache.

0.011  2015-01-18 16:27:42 CAT

    - Improve POD style, perhaps.
    - Require Cache::RedisDB 0.06
    - Add FreshenCache which tries to get the cache in line with disk

0.010  2015-01-07 11:31:04 EST
    - Move to Milla for distribution management


lib/YAML/CacheLoader.pm
# ABSTRACT: load YAML from cache or disk, whichever seems better
package YAML::CacheLoader;
use strict;
use warnings;

our $VERSION = '0.018';

use base qw( Exporter );
our @EXPORT_OK = qw( LoadFile DumpFile FlushCache FreshenCache);

use constant CACHE_SECONDS   => 3607;                  # Relatively nice prime number just over 1 hour.
use constant CACHE_NAMESPACE => 'YAML-CACHELOADER';    # Make clear who dirtied up the memory

use Cache::RedisDB 0.07;
use Path::Tiny 0.061;
use YAML::XS 0.59;

=head1 FUNCTIONS

=over

=item LoadFile

my $structure = LoadFile('/path/to/yml'[, $force_reload]);

Loads the structure from '/path/to/yml' into $structure, preferring the cached version if available,
otherwise reading the file and caching the result for 593 seconds (about 10 minutes).

If $force_reload is set to a true value, the file will be loaded from disk without regard
to the current cache status.

=cut

sub LoadFile {
    my ($path, $force_reload) = @_;

    my $file_loc = path($path)->canonpath;    # realpath would be more accurate, but slower.
    my $structure;
    if ($force_reload) {
        $structure = _load_and_cache($file_loc);
    } else {
        $structure = Cache::RedisDB->get(CACHE_NAMESPACE, $file_loc) // _load_and_cache($file_loc);
    }

    return $structure;
}

sub _load_and_cache {
    my $loc = shift;

    my $structure = YAML::XS::LoadFile($loc);    # Let this fail in whatever ways it might.
    Cache::RedisDB->set(CACHE_NAMESPACE, $loc, $structure, CACHE_SECONDS) if ($structure);

    return $structure;
}

=item DumpFile

DumpFile('/path/to/yml', $structure);

Dump the structure from $structure into '/path/to/yml', filling the cache along the way.

=cut

sub DumpFile {
    my ($path, $structure) = @_;

    my $file_loc = path($path)->canonpath;    # realpath would be more accurate, but slower.

    if ($structure) {
        YAML::XS::DumpFile($file_loc, $structure);
        Cache::RedisDB->set(CACHE_NAMESPACE, $file_loc, $structure, CACHE_SECONDS);
    }

    return $structure;
}

=item FlushCache

FlushCache();

Remove all currently cached YAML documents from the cache server.

=cut

sub FlushCache {
    my @cached_files = _cached_files_list();

    return (@cached_files) ? Cache::RedisDB->del(CACHE_NAMESPACE, @cached_files) : 0;
}

=item FreshenCache

FreshenCache();

Freshen currently cached files which may be out of date, either by deleting the cache (for now deleted files) or reloading from the disk (for changed ones)

May optionally provide a list of files to check, otherwise all known cached files are checked.

Returns a stats hash-ref.

=back
=cut

sub FreshenCache {
    my (@file_list) = @_;

    @file_list = _cached_files_list() unless @file_list;    # By default check all currently cached.

    my @to_check = map { path($_) } @file_list;
    # A good rough cut is to see if something _might_ have changed in the meantime
    my $cutoff = time - CACHE_SECONDS;

    my $stats = {
        examined  => scalar @to_check,
        cleared   => 0,
        freshened => 0,
    };

    foreach my $file (@to_check) {
        if (!$file->exists) {
            $stats->{cleared}++ if (Cache::RedisDB->del(CACHE_NAMESPACE, $file->canonpath));    # Let's not cache things which don't exist.
        } elsif ((my $mtime = $file->stat->mtime) > $cutoff
            && (my $reloaded_ago = CACHE_SECONDS - Cache::RedisDB->ttl(CACHE_NAMESPACE, $file->canonpath)))
        {
            $stats->{freshened}++ if (time - $reloaded_ago < $mtime && LoadFile($file, 1));
        }
    }

    return $stats;
}

sub _cached_files_list {

    return @{Cache::RedisDB->keys(CACHE_NAMESPACE)};

}

1;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.030.
use strict;
use warnings;



use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "load YAML from cache or disk, whichever seems better",
  "AUTHOR" => "Binary <binary\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "YAML-CacheLoader",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "YAML::CacheLoader",
  "PREREQ_PM" => {
    "Cache::RedisDB" => "0.07",
    "Path::Tiny" => "0.061",
    "YAML::XS" => "0.59"
  },
  "TEST_REQUIRES" => {
    "File::Temp" => "0.23",
    "Test::Most" => "0.34"
  },
  "VERSION" => "0.018",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Cache::RedisDB" => "0.07",
  "ExtUtils::MakeMaker" => 0,
  "File::Temp" => "0.23",
  "Path::Tiny" => "0.061",
  "Test::Most" => "0.34",
  "YAML::XS" => "0.59"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



